src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java

```diff
@@ -60,25 +62,28 @@ public void rebuildSessionFromFrame(MenuSession menuSession) throws CommCareSess
                 EntityScreen entityScreen = (EntityScreen) screen;
                 entityScreen.init(menuSession.getSessionWrapper());
                 if (entityScreen.shouldBeSkipped()) {
-                    screen = menuSession.getNextScreen();
+                    screen = menuSession.getNextScreen(false);
                     continue;
                 }
                 SessionDatum neededDatum = entityScreen.getSession().getNeededDatum();
-                Set<String> entityIds = entityScreen.getReferenceMap().keySet();
```

Comment:
There is one downside to this change to note, although I don't think it's important enough to change your approach: Calculating the reference map upfront had one positive quality which is that if there were issues with any references on the screen, they would be caught reliably and predictably. With the "minimum effort" approach, we'll non-deterministically hit those problems. 

The only impact I can think of vis-a-vis design here is asking what the inflection point is where this optimization starts to make a difference. IE: I think the current world has positive benefits which are minor (consistency and failing fast), which would be easy to outrule if this optimization is helpful fairly quickly. If it takes really big caselists before this starts to make a big impact, though, we might want to think through how we approach.

Reply:
These are fair questions, and I think it depends a lot on usage. The largest gain from this is in cases where end of form nav goes back to the case list itself, because then none of the references are evaluated. It is believable to me that this change is still worth it if we evaluate the full map every time there is a case selection (for validation purposes), but only if there is a selection. That said, I don't know what value we get from erroring when there is a valid selection if other cases in the case list have issues, which would be the main result of that.

As far as timing, I can get some concrete numbers, but it is my understanding that the "first" reference is the longest to calculate by a significant margin, because it evaluates context used by later calculations, and as such the difference between 0 and 1 is much greater than between 1 and 100. I will say that in testing, selecting a case from the first page of the case list caused significant savings, so there are large gains for early cases, but I don't yet know exactly where that inflection point is.

Reply:
Following up on a few of your questions/points:

> The largest gain from this is in cases where end of form nav goes back to the case list itself, because then none of the references are evaluated

This 100% makes sense to me and I think is important to retain regardless of the linear scan. Pushing the check for inclusion in the screen to the screen itself will likely help with that, since it can defer evaluation of ~anything to be deferred to on-demand and cached.

> I don't know what value we get from erroring when there is a valid selection if other cases in the case list have issues, which would be the main result of that.

Currently the erroring wouldn't be related to whether the selection is valid, but rather with whether the malformed entity is before or after the valid selection in an arbitrary ordering.

Since we have an end-user programming system, I think it should be a goal of the platform to both fail-fast and be as consistent as possible in a request. If something fails some % of the time, it's better if that % is 100.

> As far as timing, I can get some concrete numbers, but it is my understanding that the "first" reference is the longest to calculate by a significant margin, because it evaluates context used by later calculations, and as such the difference between 0 and 1 is much greater than between 1 and 100

This should be true in most cases for sure, as the initial reference evaluation unrolling, and the subsequent "bulk load" that fetches and inflates the records are all big bulk operations. I was curious bout the inflection because on some of the bigger county domains we did actually cross the point where it was the subsequent query/transform actions that overtook that runtime, but those were also with multi-column entities rather than just the index value expression, so I wasn't sure what % of the time to expect the linear inflate/transform part of the process to take.